### PR TITLE
Worker versioning

### PR DESCRIPTION
**Pull request by Sushisource**
Support for the new Worker Versioning feature

### CODE REVIEW

**Comment by Sushisource** [comment]
Still need to add some docstrings & experimental tags

File: packages/client/src/build-id-types.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  | MergeSets;
+
+/**
+ * This operation adds a new Build Id into a new set, which will be used as the default set for
+ * the queue. This means all new workflows will start on this Build Id.
+ *
+ * @experimental
+ */
+interface NewIdInNewDefaultSet {
+  operation: 'NEW_ID_IN_NEW_DEFAULT_SET';
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:22]: Looks like for schedules we made the action enum camel case. Let's do that here too.

File: packages/client/src/build-id-types.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  buildId: string;
+}
+
+/**
+ * This operation adds a new Build Id into an existing compatible set. The newly added ID becomes
+ * the default for that compatible set, and thus new workflow tasks for workflows which have been
+ * executing on workers in that set will now start on this new Build ID.
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:29]: Let's consistently use `Id` please.
```suggestion
 * executing on workers in that set will now start on this new Build Id.
```

File: packages/client/src/build-id-types.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ *
+ * @experimental
+ */
+interface NewCompatibleVersion {
+  operation: 'NEW_COMPATIBLE_VERSION';
+  // buildId The Build Id to add to an existing compatible set.
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:35]:
```suggestion
  // The Build Id to add to an existing compatible set.
```

Comment (bergundy) [reply packages/client/src/build-id-types.ts:35]: Or maybe `A build Id ...`

File: packages/client/src/build-id-types.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  buildId: string;
+  // A Build Id which must already be defined on the task queue, and is used to
+  // find the compatible set to add the new ID to.
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:38]:
```suggestion
  // find the compatible set to add the new Id to.
```

File: packages/client/src/build-id-types.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  existingCompatibleBuildId: string;
+  // If set to true, the targeted set will also be promoted to become the
+  // overall default set for the queue.
+  makeSetDefault?: boolean;
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:42]: Ambiguous
```suggestion
  promoteSet?: boolean;
```

File: packages/client/src/build-id-types.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+/**
+ * This operation promotes a set of compatible Build IDs to become the current
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:46]:
```suggestion
 * This operation promotes a set of compatible Build Ids to become the current
```

Comment (bergundy) [reply packages/client/src/build-id-types.ts:46]: We haven't been consistent in the reset of the codebase, looks like we use `Id` in schedules docs. And a mix of them in workflow options :(

File: packages/client/src/build-id-types.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+ * default set for the task queue. Any Build ID in the set may be used to
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:47]:
```suggestion
 * default set for the task queue. Any Build Id in the set may be used to
```

File: packages/client/src/build-id-types.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+ *
+ * @experimental
+ */
+interface PromoteSetByBuildId {
+  operation: 'PROMOTE_SET_BY_BUILD_ID';
+  buildId: string;
+}
+
+/**
+ * This operation promotes a Build ID within an existing set to become the
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:58]: Do we have to start all of the doc strings with "This operation"? Seems redundant.

File: packages/client/src/build-id-types.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+ * default ID for that set.
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:59]:
```suggestion
 * default Id for that set.
```

File: packages/client/src/build-id-types.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+interface MergeSets {
+  operation: 'MERGE_SETS';
+  // A Build Id which is used to find the primary set to be merged.
+  primaryBuildId: string;
+  // A Build Id which is used to find the secondary set to be merged.
+  secondaryBuildId: string;
+}
+
+/**
+ * Represents the sets of compatible Build ID versions associated with some
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:84]:
```suggestion
 * Represents the sets of compatible Build Id versions associated with some
```

File: packages/client/src/build-id-types.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+ * Task Queue, as fetched by {@link WorkflowClient.getWorkerBuildIdCompatability}.
+ *
+ * @experimental
+ */
+export class WorkerBuildIdVersionSets {
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:89]: We don't return classes for simple objects like this all over the code base.
Instead define two interfaces:

```
export interface TaskQueueMetadata {
  versionSets: VersionSet[];
  defaultSet: string;
}

export interface VersionSet {
  buildIds: string[];
  defaultBuildId: string;
}
```

I regret that we didn't make the API return all of the task queue metadata `TaskQueueMetadata` as we're going to add more information to it.

For now you can name it `WorkerVersioningInfo`

File: packages/client/src/workflow-client.ts [hunk 12]
```diff
@@ -69,6 +69,9 @@ import {
   WithDefaults,
 } from './base-client';
 import { mapAsyncIterable } from './iterators-utils';
+import { BuildIdOperation, WorkerBuildIdVersionSets } from './build-id-types';
+import IUpdateWorkerBuildIdCompatibilityRequest = temporal.api.workflowservice.v1.IUpdateWorkerBuildIdCompatibilityRequest;
+import { assertNever } from '@temporalio/common/lib/type-helpers';
```
Comment (bergundy) [inline packages/client/src/workflow-client.ts:74]: lint

File: packages/client/src/workflow-client.ts [hunk 13]
```diff
@@ -581,6 +591,69 @@ export class WorkflowClient extends BaseClient {
     }
   }
 
+  /**
+   * Used to add new Build IDs or otherwise update the relative compatibility of Build IDs as
```
Comment (bergundy) [inline packages/client/src/workflow-client.ts:595]:
```suggestion
   * Used to add new Build Ids or otherwise update the relative compatibility of Build Ids as
```

File: packages/client/src/workflow-client.ts [hunk 14]
```diff
@@ -581,6 +591,69 @@ export class WorkflowClient extends BaseClient {
...
+   * defined on a specific task queue for the Worker Versioning feature. For more on this feature,
+   * see https://docs.temporal.io/workers#worker-versioning
+   *
+   * @param taskQueue The task queue to make changes to.
+   * @param operation The operation to be performed.
+   *
+   * @experimental
+   */
+  public async updateWorkerBuildIdCompatability(taskQueue: string, operation: BuildIdOperation): Promise<void> {
```
Comment (bergundy) [inline packages/client/src/workflow-client.ts:604]: Make a new TaskQueueClient and put these methods on it and attach it to the top level `Client` instead.

`client.taskQueue.updateBuildIdCompatibility`

Comment (bergundy) [inline packages/client/src/workflow-client.ts:604]: spelling: compatibility

File: packages/common/src/type-helpers.ts [hunk 15]
```diff
@@ -61,3 +61,10 @@ export function errorCode(error: unknown): string | undefined {
 
   return undefined;
 }
+
+/**
+ * Asserts that some type is the never type
+ */
+export function assertNever(x: never): never {
```
Comment (bergundy) [inline packages/common/src/type-helpers.ts:68]: Make a specific helper for this use case and throw a TypeError.

File: packages/client/src/workflow-client.ts [hunk 16]
```diff
@@ -487,7 +490,14 @@ export class WorkflowClient extends BaseClient {
       try {
         res = await this.workflowService.getWorkflowExecutionHistory(req);
       } catch (err) {
-        this.rethrowGrpcError(err, { workflowId, runId }, 'Failed to get Workflow execution history');
+        this.rethrowGrpcError(
```
Comment (bergundy) [inline packages/client/src/workflow-client.ts:493]: Why?

File: packages/client/src/workflow-client.ts (hunk 14 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:604]: The word worker will be redundant with a separate client.
Also need a high level API for reachability, put that on the new client.

File: packages/core-bridge/sdk-core
Comment (bergundy) [inline packages/core-bridge/sdk-core:1]: Do we need to update core for this PR?

File: packages/worker/src/worker-options.ts [hunk 17]
```diff
@@ -81,12 +81,26 @@ export interface WorkerOptions {
   /**
    * A string that should be unique to the exact worker code/binary being executed.
    *
-   * This is used to populate the `binaryChecksum` attribute in history events originated from this Worker.
```
Comment (bergundy) [inline packages/worker/src/worker-options.ts:84]: Please mention that this will be used for binaryChecksum in older servers.

File: packages/worker/src/worker-options.ts [hunk 18]
```diff
@@ -81,12 +81,26 @@ export interface WorkerOptions {
...
    *
    * @default `@temporalio/worker` package name and version + checksum of workflow bundle's code
+   *
+   * @experimental
    */
   buildId?: string;
 
+  /**
+   * If set true, this worker opts into the worker versioning feature. This ensures it only receives
+   * workflow tasks for workflows which it claims to be compatible with.
```
Comment (bergundy) [inline packages/worker/src/worker-options.ts:96]: I'd add a sentence that this affects how we use the buildId option above.

File: packages/client/src/build-id-types.ts (hunk 3 above)
Comment (Sushisource) [reply packages/client/src/build-id-types.ts:35]: Yeah this is just a copy paste mistake

File: packages/client/src/build-id-types.ts (hunk 1 above)
Comment (Sushisource) [reply packages/client/src/build-id-types.ts:22]: You led me astray in the airport!

File: packages/client/src/build-id-types.ts (hunk 11 above)
Comment (Sushisource) [reply packages/client/src/build-id-types.ts:89]: I'm not sure we should rename them as it's consistent with what I did in the other SDKs right now, but that's not the biggest deal. If you really prefer that we can rename

File: packages/common/src/type-helpers.ts (hunk 15 above)
Comment (Sushisource) [reply packages/common/src/type-helpers.ts:68]: What do you mean specific helper? This will work for all cases where you want to do exhausive checks, pretty useful. The throw is actually irrelevant, since what matters happens at compile time, but I changed it anyway.

File: packages/client/src/workflow-client.ts (hunk 16 above)
Comment (Sushisource) [reply packages/client/src/workflow-client.ts:493]: Just got autoformatted like that. Not sure how it wasn't before.

File: packages/core-bridge/sdk-core
Comment (Sushisource) [reply packages/core-bridge/sdk-core:1]: It already has been

File: packages/client/src/workflow-client.ts (hunk 14 above)
Comment (Sushisource) [reply packages/client/src/workflow-client.ts]: I'm fine adding the reachability client but we didn't in Go/Java since there just wasn't a whole lot to abstract over. It can probably be a little nicer in TS though.

File: packages/test/src/test-worker-versioning.ts [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+      operation: 'mergeSets',
+      primaryBuildId: '2.0',
+      secondaryBuildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '2.0');
+
+    const reachResp = await conn.taskQueue.getBuildIdReachability({ buildIds: ['2.0', '1.0', '1.1'] });
+    assert.deepEqual(reachResp.buildIdReachability.get('2.0')?.taskQueueReachability.get(taskQueue), ['NewWorkflows']);
+    assert.deepEqual(reachResp.buildIdReachability.get('1.1')?.taskQueueReachability.get(taskQueue), []);
```
Comment (bergundy) [inline packages/test/src/test-worker-versioning.ts:176]: This will fail once we add a grace time of 5 minutes for the reachability.

File: packages/test/src/test-worker-versioning.ts [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'mergeSets',
+      primaryBuildId: '2.0',
+      secondaryBuildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '2.0');
+
+    const reachResp = await conn.taskQueue.getBuildIdReachability({ buildIds: ['2.0', '1.0', '1.1'] });
```
Comment (bergundy) [inline packages/test/src/test-worker-versioning.ts:174]: This whole part should run in an "eventually" block (retry for a while) because of eventual consistency.

File: packages/test/src/test-worker-versioning.ts [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+    assert.equal(resp?.defaultBuildId(), '1.1');
+
+    // Target nonexistent build ID
+    await t.throwsAsync(
+      conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+        operation: 'newCompatibleVersion',
+        buildId: '1.2',
+        existingCompatibleBuildId: 'amnotreal',
+      }),
+      { message: /amnotreal not found/ }
```
Comment (bergundy) [inline packages/test/src/test-worker-versioning.ts:142]: What's the error class that's thrown here?

File: packages/client/src/task-queue-client.ts [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+  /** The kind of reachability this request is concerned with. */
+  reachability?: ReachabilityType;
+}
+
+export interface ReachabilityResponse {
+  /** Maps Build Ids to their reachability information. */
+  buildIdReachability: Map<string, BuildIdReachability>;
+}
+
+type ReachabilityTypeResponse = ReachabilityType | 'NotFetched';
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/client/src/task-queue-client.ts:161]: That rename doesn't make sense to me, it's not the response type - it's the type *in the response*. Keeping unspecified also seems sort of a loss since the whole point of the higher level API is to make things more clear, but unspecified is not clear.

File: packages/client/src/task-queue-client.ts [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+   *  concerned with.
+   */
+  taskQueues?: string[];
+  /** The kind of reachability this request is concerned with. */
+  reachability?: ReachabilityType;
+}
+
+export interface ReachabilityResponse {
+  /** Maps Build Ids to their reachability information. */
+  buildIdReachability: Map<string, BuildIdReachability>;
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/client/src/task-queue-client.ts:158]: Record works fine

File: packages/client/src/workflow-client.ts (hunk 16 above)
Comment (Sushisource) [reply packages/client/src/workflow-client.ts:490]: Yeah but the CI formatter doesn't correct it. :shrug:

File: packages/common/src/type-helpers.ts (hunk 15 above)
Comment (Sushisource) [reply packages/common/src/type-helpers.ts:73]: Any extension of the enum will render this as a compile error. The throw could only happen I believe if you were doing weird JS things and going around TS.

File: packages/client/src/build-id-types.ts [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+   * Returns the default set of compatible Build Ids for the task queue these sets are
+   * associated with.
+   */
+  defaultSet(): BuildIdVersionSet;
+
+  /**
+   * Returns the overall default Build Id for the task queue these sets are
+   * associated with.
+   */
+  defaultBuildId(): string;
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/client/src/build-id-types.ts:105]: Cool, I forgot getters exist. IMO it's... kinda wrong, since it hides arbitrary cost operations behind what looks like pure memory access, but, it's nice for interface compatibility.

File: packages/core-bridge/sdk-core
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/core-bridge/sdk-core:1]: It's not unrelated, it's needed.

File: packages/common/src/type-helpers.ts (hunk 15 above)
Comment (bergundy) [reply packages/common/src/type-helpers.ts:73]: No, the server may be updated with new enum values and the SDK may not have those.

File: packages/test/src/test-worker-versioning.ts (hunk 21 above)
Comment (Sushisource) [reply packages/test/src/test-worker-versioning.ts:142]: I think it's just `Error`:

```
Error {
    code: 5,
    details: 'targeted compatible_version amnotreal not found',
    metadata: Metadata {
      internalRepr: Map {
        'content-type' => Array [ … ],
        'grpc-status-details-bin' => Array [ … ],
      },
      options: {},
    },
    message: '5 NOT_FOUND: targeted compatible_version amnotreal not found',
  }
  ```

File: packages/test/src/test-worker-versioning.ts (hunk 19 above)
Comment (Sushisource) [reply packages/test/src/test-worker-versioning.ts:176]: I can just assert that the unreachable ones are present and not bother with them being empty? We don't really need to test that server is sending the right thing back, just that we give empty when they aren't reachable.

Other option is just unit test with fake data.

File: packages/test/src/test-worker-versioning.ts (hunk 21 above)
Comment (bergundy) [reply packages/test/src/test-worker-versioning.ts:144]: You need to wrap the call in `rethrowGrpcError` like we do in the other clients.

Comment (Sushisource) [reply packages/test/src/test-worker-versioning.ts:144]: K. I added not found but it's not really possible to tie one specific build id to the error in all cases, but that seems fine since the error message itself will have that info and the main thing is matching on type.

File: packages/common/src/type-helpers.ts (hunk 15 above)
Comment (Sushisource) [reply packages/common/src/type-helpers.ts:73]: Sure, makes sense.

File: packages/client/src/build-id-types.ts [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+ * Represents one set of compatible Build Ids.
+ *
+ * @experimental
+ */
+export interface BuildIdVersionSet {
+  // All build IDs contained in the set.
+  readonly buildIds: string[];
+
+  // Returns the default Build Id for this set
+  default(): string;
```
Comment (bergundy) [inline packages/client/src/build-id-types.ts:118]: Same with the other interface, make it an attribute not a method.

File: packages/client/src/index.ts [hunk 26]
```diff
@@ -38,3 +38,4 @@ export * from './workflow-client';
 export * from './workflow-options';
 export * from './schedule-types';
 export * from './schedule-client';
+export { WorkerBuildIdVersionSets, BuildIdVersionSet, BuildIdOperation } from './build-id-types';
```
Comment (bergundy) [inline packages/client/src/index.ts:41]:
```suggestion
export * from './task-queue-client';
export { WorkerBuildIdVersionSets, BuildIdVersionSet, BuildIdOperation } from './build-id-types';
```

File: packages/client/src/workflow-client.ts (hunk 16 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:490]: Can you please reset this file, there's no relevant change here.

File: packages/client/src/index.ts (hunk 26 above)
Comment (Sushisource) [reply packages/client/src/index.ts:41]: We don't actually want to export everything though

File: packages/client/src/task-queue-client.ts [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+        reachability: reachabilityTypeToProto(options.reachability),
+      });
+    } catch (e) {
+      this.rethrowGrpcError(e, 'Unexpected error fetching Build Id reachability');
+    }
+    return reachabilityResponseFromProto(resp);
+  }
+
+  protected rethrowGrpcError(err: unknown, fallbackMessage: string): never {
+    if (isServerErrorResponse(err)) {
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:144]: See how James changed this method in workflow-client:

```
    if (isGrpcServiceError(err)) {
      rethrowKnownErrorTypes(err);
```

File: packages/client/src/task-queue-client.ts [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+/**
+ * There are different types of reachability:
+ *   - `NewWorkflows`: The Build Id might be used by new workflows
+ *   - `ExistingWorkflows` The Build Id might be used by open workflows and/or closed workflows.
+ *   - `OpenWorkflows` The Build Id might be used by open workflows
+ *   - `ClosedWorkflows` The Build Id might be used by closed workflows
+ */
+export type ReachabilityType = 'NewWorkflows' | 'ExistingWorkflows' | 'OpenWorkflows' | 'ClosedWorkflows';
+
+interface BaseReachabilityOptions {
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:168]: Missing export
```suggestion
export interface BaseReachabilityOptions {
```

File: packages/client/src/task-queue-client.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply packages/client/src/task-queue-client.ts:188]: Still need to export this

File: packages/client/src/task-queue-client.ts (hunk 28 above)
Comment (Sushisource) [reply packages/client/src/task-queue-client.ts:168]: This doesn't actually get used directly by users, only through the `ReachabilityOptions` type

File: packages/client/src/task-queue-client.ts [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+          continue;
+        }
+        if (tqr.reachability == null) {
+          taskQueueReachability[tqr.taskQueue] = [];
+          continue;
+        }
+        taskQueueReachability[tqr.taskQueue] = tqr.reachability.map(reachabilityTypeFromProto);
+      }
+    }
+    buildIdReachability[bir.buildId ?? ''] = { taskQueueReachability };
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:232]: I mentioned this before, using empty strings for keys is weird, let's make a symbol for the unversioned build ID (and please export it).

```ts
export const UnversionedBuildId = Symbol.for('__temporal_unversionedBuildId')
```

This function will return 
```
 { buildIdReachabilty: Record<string | UnversionedBuildId, BuildIdReachability> }
```

And in the request you'll accept
```
buildIds: Array<string | UnversionedBuildId>
```

File: packages/common/src/type-helpers.ts (hunk 15 above)
Comment (bergundy) [reply packages/common/src/type-helpers.ts:73]: I guess my only request would be to pass in the error message to this function so it doesn't just say `Unexpected object` which isn't very helpful especially with enums.

File: packages/client/src/task-queue-client.ts [hunk 30]
```diff
@@ -0,0 +1,297 @@
...
+        break;
+      default:
+        assertNever(operation);
+    }
+    try {
+      await this.workflowService.updateWorkerBuildIdCompatibility(request);
+    } catch (e) {
+      if (isGrpcServiceError(e)) {
+        rethrowKnownErrorTypes(e);
+      }
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:100]: Put this in `rethrowGrpcError` to avoid repetition.

File: packages/client/src/task-queue-client.ts [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+    let resp;
+    try {
+      resp = await this.workflowService.getWorkerBuildIdCompatibility({
+        taskQueue,
+        namespace: this.options.namespace,
+      });
+    } catch (e) {
+      if (isGrpcServiceError(e)) {
+        rethrowKnownErrorTypes(e);
+      }
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:122]: Same

File: packages/client/src/task-queue-client.ts [hunk 32]
```diff
@@ -0,0 +1,297 @@
...
+      resp = await this.workflowService.getWorkerTaskReachability({
+        namespace: this.options.namespace,
+        taskQueues: options.taskQueues,
+        buildIds,
+        reachability: reachabilityTypeToProto(options.reachability),
+      });
+    } catch (e) {
+      if (isGrpcServiceError(e)) {
+        rethrowKnownErrorTypes(e);
+      }
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:160]: Same, avoid repetition.

File: packages/client/src/task-queue-client.ts [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+ */
+export interface BaseReachabilityOptions {
+  /**
+   * A list of build ids to query the reachability of. Currently, at least one Build Id must be
+   * specified, but this restriction may be lifted in the future.
+   */
+  buildIds: (string | UnversionedBuildIdType)[];
+  /**
+   *  A list of task queues with Build Ids defined on them that the request is
+   *  concerned with.
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:202]:
```suggestion
   * A list of task queues with Build Ids defined on them that the request is
   * concerned with.
```

File: packages/client/src/task-queue-client.ts [hunk 34]
```diff
@@ -0,0 +1,297 @@
...
+ * @experimental
+ */
+export type LoadedTaskQueueClientOptions = LoadedWithDefaults<TaskQueueClientOptions>;
+
+/**
+ * A stand-in for a Build Id for unversioned Workers
+ * @experimental
+ */
+export const UnversionedBuildId = Symbol.for('__temporal_unversionedBuildId');
+type UnversionedBuildIdType = typeof UnversionedBuildId;
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:28]:
```suggestion
export type UnversionedBuildIdType = typeof UnversionedBuildId;
```

File: packages/client/src/task-queue-client.ts [hunk 35]
```diff
@@ -0,0 +1,297 @@
...
+  /** The kind of reachability this request is concerned with. */
+  reachability?: ReachabilityType;
+}
+
+export interface ReachabilityResponse {
+  /** Maps Build Ids to their reachability information. */
+  buildIdReachability: Record<string | UnversionedBuildIdType, BuildIdReachability>;
+}
+
+type ReachabilityTypeResponse = ReachabilityType | 'NotFetched';
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:214]:
```suggestion
export type ReachabilityTypeResponse = ReachabilityType | 'NotFetched';
```

File: packages/client/src/task-queue-client.ts [hunk 36]
```diff
@@ -0,0 +1,297 @@
...
+    case 'NewWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_NEW_WORKFLOWS;
+    case 'ExistingWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_EXISTING_WORKFLOWS;
+    case 'OpenWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_OPEN_WORKFLOWS;
+    case 'ClosedWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_CLOSED_WORKFLOWS;
+    default:
+      assertNever(type);
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:238]:
```suggestion
      assertNever(type, 'Unknown reachability type');
```

File: packages/client/src/task-queue-client.ts [hunk 37]
```diff
@@ -0,0 +1,297 @@
...
+    case temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_NEW_WORKFLOWS:
+      return 'NewWorkflows';
+    case temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_EXISTING_WORKFLOWS:
+      return 'ExistingWorkflows';
+    case temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_OPEN_WORKFLOWS:
+      return 'OpenWorkflows';
+    case temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_CLOSED_WORKFLOWS:
+      return 'ClosedWorkflows';
+    default:
+      return assertNever(rtype);
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:278]:
```suggestion
      return assertNever(rtype, 'Unknown reachability type');
```

File: packages/client/src/task-queue-client.ts [hunk 38]
```diff
@@ -0,0 +1,297 @@
...
+ * It could be because:
+ * - Id passed is incorrect
+ * - Build Id has been scavenged by the server.
+ *
+ * @experimental
+ */
+export class BuildIdNotFoundError extends Error {
+  public readonly name: string = 'BuildIdNotFoundError';
+
+  constructor(message: string) {
```
Comment (bergundy) [inline packages/client/src/task-queue-client.ts:294]: If you're not tired from me discovering things every round, you could specify which build id is not found in this exception, could have a `buildId: string` attribute on this class.

File: packages/client/src/workflow-client.ts
Comment (bergundy) [inline packages/client/src/workflow-client.ts:1]: nit: No need to change this file, please reset.

File: packages/test/src/integration-tests.ts
Comment (bergundy) [inline packages/test/src/integration-tests.ts:1]: Reset changes to this file please.

File: packages/test/src/test-worker-versioning-unit.ts
Comment (bergundy) [inline packages/test/src/test-worker-versioning-unit.ts:1]: nit: merge into the other test file.

File: packages/test/src/test-worker-versioning.ts (hunk 21 above)
Comment (bergundy) [reply packages/test/src/test-worker-versioning.ts:146]: Add the type to the assertion.

File: packages/test/src/test-worker-versioning.ts [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+      operation: 'mergeSets',
+      primaryBuildId: '2.0',
+      secondaryBuildId: '1.0',
+    });
+    resp = await conn.taskQueue.getBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId, '2.0');
+
+    await asyncRetry(
+      async () => {
+        const reachResp = await conn.taskQueue.getBuildIdReachability({ buildIds: ['2.0', '1.0', '1.1'] });
```
Comment (bergundy) [inline packages/test/src/test-worker-versioning.ts:180]: Can you please add a test for reachability of the unversioned build id?

File: packages/client/src/task-queue-client.ts (hunk 38 above)
Comment (Sushisource) [reply packages/client/src/task-queue-client.ts:294]: You can't actually do that, because in the merge operation it could be either of two and the error isn't readily parseable. Mentioned here: https://github.com/temporalio/sdk-typescript/pull/1146#discussion_r1244088033

File: packages/test/src/test-worker-versioning-unit.ts
Comment (Sushisource) [reply packages/test/src/test-worker-versioning-unit.ts:1]: I couldn't see any obvious way to have different `test` instances in the same file

File: packages/test/src/integration-tests.ts
Comment (Sushisource) [reply packages/test/src/integration-tests.ts:1]: Why? This will work with old versions

File: packages/test/src/integration-tests.ts [hunk 40]
```diff
@@ -624,14 +634,28 @@ export function runIntegrationTests(codec?: PayloadCodec): void {
...
       CustomDoubleField: [3.14],
-      BuildIds: ['unversioned', `unversioned:${t.context.worker.options.buildId}`],
     });
+    let checksum: any;
+    if (BinaryChecksums != null) {
+      t.true(BinaryChecksums.length === 1);
+      checksum = BinaryChecksums[0];
+    } else {
+      t.true(BuildIds!.length === 2);
+      t.deepEqual(BuildIds![0], 'unversioned');
```
Comment (bergundy) [inline packages/test/src/integration-tests.ts:651]: nit: just deepEquals on the BuildIds.

**Review by bergundy — APPROVED** [review]
